### PR TITLE
Certification test for Azure CosmosDB Statestore

### PR DESCRIPTION
**Pull request by shivamkm07**
Signed-off-by: shivam <shivamkm07@gmail.com>

# Description

Certification test for Azure CosmosDB statestore

## Issue reference

Please reference the issue this PR will close: #1138 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: tests/certification/state/azure/cosmosdb/cosmosdb_test.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+				err = client.SaveState(ctx, statestore, stateKey, []byte(stateValue), setMeta)
+				assert.NoError(t, err)
+
+				// get state
+				item, err := client.GetState(ctx, statestore, stateKey, getMeta)
+				assert.NoError(t, err)
+				assert.Equal(t, expectedValue, string(item.Value))
+
+				// delete state
+				err = client.DeleteState(ctx, statestore, stateKey, setMeta)
```
Comment (DeepanshuA) [inline tests/certification/state/azure/cosmosdb/cosmosdb_test.go:111]: Should we also try to delete with wrong partition key and assert error?

File: tests/certification/state/azure/cosmosdb/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Azure Cosmos DB certification testing
+
+This project aims to test the Azure Cosmos DB State Store component under various conditions.
+
+## Test plan
+
+### Basic Test using master key authentication
```
Comment (berndverst) [inline tests/certification/state/azure/cosmosdb/README.md:7]: Please remove this. It's already covered by conformance tests.

File: tests/certification/state/azure/cosmosdb/README.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+1. Able to create and test connection.
+2. Able to do set, fetch and delete.
+
+### Test Partition Keys
+Ensure the following scenarios:
+ 1. In case of invalid partition key, the access fails
```
Comment (berndverst) [inline tests/certification/state/azure/cosmosdb/README.md:13]: Access? That sounds like it's about authorization.. but that's not so.

Please be specific that this is about a Get or Set operation.

File: tests/certification/state/azure/cosmosdb/README.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ 2. In case of missing partition key, *key* property is used in place of it
```
Comment (berndverst) [inline tests/certification/state/azure/cosmosdb/README.md:14]: Also check what happens if this key property is missing!

File: tests/certification/state/azure/cosmosdb/README.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+This project aims to test the Azure Cosmos DB State Store component under various conditions.
+
+## Test plan
+
+### Basic Test using master key authentication
+1. Able to create and test connection.
+2. Able to do set, fetch and delete.
+
+### Test Partition Keys
```
Comment (berndverst) [inline tests/certification/state/azure/cosmosdb/README.md:11]: Please add a test where you are connecting to a different collection -- a collection which does not use the expected `/partitionKey` partition key. We want to assert the failure for that case. This is a common problem.

To do so, please add a new collection with wrong partition name here:
https://github.com/dapr/components-contrib/blob/master/.github/infrastructure/conformance/azure/conf-test-azure-cosmosdb.bicep#L44

Then also make sure to update the following so that the second collection name gets exported to the KeyVault where we save these configurations
https://github.com/dapr/components-contrib/blob/master/.github/infrastructure/conformance/azure/conf-test-azure.bicep
https://github.com/dapr/components-contrib/blob/master/.github/infrastructure/conformance/azure/setup-azure-conf-test.sh

Lastly, set up a new component YAML talking to the collection with the bad partition name. And make sure to reference the new environment variable you are creating above.

Don't forget to inject that environment variable into the workflow run via the required-secrets section!

**Review by berndverst — CHANGES_REQUESTED** [review]

**Review by berndverst — CHANGES_REQUESTED** [review]
Please also add tests that check the handling of null data, empty string - especially with different content types specified.

https://github.com/dapr/components-contrib/pull/1974

**Comment by mukundansundar** [comment]
> Please also add tests that check the handling of null data, empty string - especially with different content types specified.
> 
> #1974

@berndverst should the nil, empty string test be part of conformance tests?

**Comment by DeepanshuA** [comment]
@berndverst Please advise, does it make sense to include this test as well? : 
https://github.com/dapr/components-contrib/issues/1138#issuecomment-1217564063

**Comment by berndverst** [comment]
> @berndverst Please advise, does it make sense to include this test as well? : [#1138 (comment)](https://github.com/dapr/components-contrib/issues/1138#issuecomment-1217564063)

Ok to skip, I am going to add it here: https://github.com/dapr/components-contrib/pull/1992

File: tests/certification/state/azure/cosmosdb/README.md (hunk 5 above)
Comment (berndverst) [reply tests/certification/state/azure/cosmosdb/README.md:11]: To provide more context -- what you have described is only impacting partition key values, but the Cosmos DB state store requires that the Cosmos DB collection was setup with a partition Key that is exactly called `/partitionKey`. It will be good to assert the failure when the collection has been set up wrong.

**Review by berndverst — CHANGES_REQUESTED** [review]
@shivamkm07 please address review comments.

File: tests/certification/state/azure/cosmosdb/cosmosdb_test.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	log := logger.NewLogger("dapr.components")
+	ports, err := dapr_testing.GetFreePorts(2)
+	assert.NoError(t, err)
+
+	currentGrpcPort := ports[0]
+	currentHTTPPort := ports[1]
+
+	basicTest := func(statestore string) flow.Runnable {
+		return func(ctx flow.Context) error {
+			client, err := client.NewClientWithPort(fmt.Sprint(currentGrpcPort))
```
Comment (ItalyPaleAle) [inline tests/certification/state/azure/cosmosdb/cosmosdb_test.go:53]: What's the purpose of the fmt.Sprintf here?

File: tests/certification/state/azure/cosmosdb/go.mod [hunk 7]
```diff
@@ -0,0 +1,146 @@
+module github.com/dapr/components-contrib/tests/certification/state/azure/cosmosdb
+
+go 1.18
```
Comment (ItalyPaleAle) [inline tests/certification/state/azure/cosmosdb/go.mod:3]: 1.19

File: tests/certification/state/azure/cosmosdb/go.mod [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	k8s.io/klog/v2 v2.30.0 // indirect
+	k8s.io/kube-openapi v0.0.0-20211115234752-e816edb12b65 // indirect
+	k8s.io/utils v0.0.0-20210930125809-cb0fa318a74b // indirect
+	sigs.k8s.io/controller-runtime v0.11.0 // indirect
+	sigs.k8s.io/json v0.0.0-20211020170558-c049b76a60c6 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.0 // indirect
+	sigs.k8s.io/yaml v1.3.0 // indirect
+)
+
+replace go.opentelemetry.io/otel => go.opentelemetry.io/otel v0.20.0
```
Comment (ItalyPaleAle) [inline tests/certification/state/azure/cosmosdb/go.mod:142]: I think this needs to be removed with the newer Dapr versions

Comment (berndverst) [reply tests/certification/state/azure/cosmosdb/go.mod:142]: Correct. This needs to be removed.

**Review by berndverst — APPROVED** [review]
We depend on this PR to upgrade the Track2 SDK. Therefore I will merge this now -- I fixed this up as necessary.
